core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -618,6 +625,230 @@ public void run() {
             }
         }
 
+        public void handleLogStartOffsetUpdate(TopicPartition topicPartition, long remoteLogStartOffset) {
+            if (isLeader()) {
```

Comment:
even if we are not the leader at this stage, we have deleted the logs in remote. Shouldn't we still update the metadata?

Reply:
I agree with @divijvaidya 's suggestion. In this scenario:
1. replica 1 is the leader, and doing remote log segment deletion
2. leadership changed to replica 2
3. replica 1 entering this `handleLogStartOffsetUpdate` method

Under current implementation, we won't update log start offset since it is not the leader anymore. But we should update it! @satishd , thoughts?

Reply:
It is not mandatory to update it when this node becomes a follower as the existing follower fetch protocol makes sure that the follower truncates their log-start-offset based on the leader's log-start-ffset.

Reply:
So you mean, all the segment deletion will happen again in the new leader, and update the log start offset there. OK, make sense.

Reply:
In the FETCH response, the leader-log-start-offset will be piggy-backed. But, there can be a scenario:

1. Leader deleted the remote log segment and updates it's log-start-offset
2. Before the replica-2 update it's log-start-offset via FETCH-request, the leadership changed to replica-2.
3. There are no more eligible segments to delete from remote.
4. The log-start-offset will be stale (referring to old log-start-offset but the data was already removed from remote)
5. If the consumer starts to read from the beginning of the topic, it will fail to read.

It's good to handle this case in this PR itself.

Reply:
Similar to local, which marks the segment for deletion (renames the file to .delete) and deletes it after 1 minute. (segment.delete.delay.ms). Should we move the log-start-offset before the remote log segment deletion?

Reply:
@kamalcph The case mentioned by you can be addressed in a followup PR. Please file a JIRA.

Reply:
Filed KAFKA-15351 and KAFKA-15352 to track the cases.